### PR TITLE
Fix bug for running workbench

### PR DESCRIPTION
**Pull request by shenli**
Fix bug in type inferrer. Add missing type info.
@coocood @qiuyesuifeng

### CODE REVIEW

File: optimizer/typeinferer.go [hunk 1]
```diff
@@ -43,34 +47,57 @@ func (v *typeInferrer) Leave(in ast.Node) (out ast.Node, ok bool) {
...
 		x.SetType(types.NewFieldType(mysql.TypeLonglong))
+		x.Type.Charset = charset.CharsetBin
+		x.Type.Collate = charset.CollationBin
 	case *ast.SelectStmt:
 		v.selectStmt(x)
 	case *ast.UnaryOperationExpr:
 		v.unaryOperation(x)
+	case *ast.ValueExpr:
+		v.handleValueExpr(x)
+	case *ast.FuncCallExpr:
```
Comment (qiuyesuifeng) [inline optimizer/typeinferer.go]: Please keep in alphabetical order.

File: util/types/field_type.go [hunk 2]
```diff
@@ -146,47 +146,74 @@ func IsDataItem(d interface{}) bool {
...
 	case mysql.Set:
-		return NewFieldType(mysql.TypeSet)
+		tp = NewFieldType(mysql.TypeSet)
+		tp.Charset = charset.CharsetBin
+		tp.Collate = charset.CharsetBin
 	case *DataItem:
-		return x.Type
+		tp = x.Type
+	default:
+		tp = NewFieldType(mysql.TypeDecimal)
```
Comment (qiuyesuifeng) [inline util/types/field_type.go]: Should we define an unknown type like `TypeUnknown`?

File: optimizer/typeinferer.go [hunk 3]
```diff
@@ -163,4 +194,77 @@ func (v *typeInferrer) unaryOperation(x *ast.UnaryOperationExpr) {
...
+		tp = types.NewFieldType(mysql.TypeLonglong)
+		tp.Flag |= mysql.UnsignedFlag
+	case "if":
+		tp = x.Args[1].GetType()
+	default:
+		// TypeDecimal means unknown type.
+		// For nullif, we do not know the type.
+		tp = types.NewFieldType(mysql.TypeDecimal)
+	}
+	if len(tp.Charset) == 0 {
```
Comment (qiuyesuifeng) [inline optimizer/typeinferer.go]: Could u show some detail for this check？

File: util/types/field_type.go (hunk 2 above)
Comment (shenli) [inline util/types/field_type.go]: TypeDecimal is not used in mysql.
We use it as uninitialized type in TiDB.

Comment (ngaut) [inline util/types/field_type.go]: Could you add some comments?

Comment (shenli) [inline util/types/field_type.go]: I will define a const with a self-explanatory name.

**Comment by shenli** [comment]
@coocood

**Comment by coocood** [comment]
LGTM

**Pull request merged**